In [3]:
a = "Hello Python"
a = a[6:12]+" "+ a[0:5]
print a

Python Hello


a="Hello Python" 문자열을 Slicing과 연결 연산자 (+)를 사용하여 "Python Hello"로 변경하시오.
[주의] a 변수 자체에 "Python Hello" 문자열이 저장되어야 함

slicing하여 필요한 두부분을 a에 다시 저장하였습니다.

In [4]:
b="Hello Python World"
b = b[13:18] + " " + b[6:12] + " " + b[0:5]
print b

World Python Hello


b="Hello Python World" 문자열을 Slicing과 연결 연산자 (+)를 사용하여 "World Python Hello"로 변경하시오.
[주의] b 변수 자체에 "Python Hello" 문자열이 저장되어야 함

1번 문제와 같이 slicing하여 필요한 두부분을 다시 저장하였습니다.

In [5]:
c = "Hello"
c = c[::-1]
print c

olleH


c="Hello"를 "olleH"로 변경하시오
[주의] c 변수 자체에 "olleH" 문자열이 저장되어야 함

slicing[::]을 사용할때 [::-1]을 하면 문자열을 뒤집어 반환합니다.

In [6]:
s="python"
print s[0], s[0][0], s[0][0][0]

p p p


[0], [0][0], [0][0][0] 모두 가장 처음의 값을 가리키기 때문에 모두 같은 p를 출력해낸다.

In [7]:
print s[-100], s[100]

IndexError: string index out of range

인덱스에 들어간 수가 범위를 벗어났습니다.

In [8]:
print s[-100, 100]

TypeError: string indices must be integers, not tuple

In [9]:
print s[1 : -1]

ytho


1(y)부터 시작하여 인덱스-1(n) 앞까지 슬라이싱 한다.

In [10]:
print s[3 : -3]

인덱스 3(h)부터 인덱스 -3(h)앞까지 슬라이싱하는데 이때 부합하는 원소가 없다.

s="python"에 대해 다음 문제를 풀어보시오.
s[0], s[0][0], s[0][0][0]은 각각 어떤 값이 나오는지 확인하고 그 이유를 나름대로 설명해 보시오.
s[-100], s[100]은 값이 나오는지 에러가 나오는지 확인하고 그 결과에 대한 이유를 나름대로 설명해 보시오.
s[-100, 100]은 값이 나오는지 에러가 나오는지 확인하고 그 결과에 대한 이유를 나름대로 설명해 보시오.
s[1:-1]의 결과를 확인하고 그 결과에 대한 이유를 정확하게 설명하시오.
s[3:-3]의 결과를 확인하고 그 결과에 대한 이유를 정확하게 설명하시오.

In [11]:
for i in range(1, 100) :
    if i % 2 == 1 :
        print i

1
3
5
7
9
11
13
15
17
19
21
23
25
27
29
31
33
35
37
39
41
43
45
47
49
51
53
55
57
59
61
63
65
67
69
71
73
75
77
79
81
83
85
87
89
91
93
95
97
99


for문을 활용하여 1부터 100사이의 홀수를 출력하시오.

c와 다르게 for문에 in range()를 사용하여 범위를 지정한후에 2로 나눈 나머지값이 1이 되는 수들을 출력한다.

In [12]:
result = 0
i = 0
while i <= 100:
    if i % 2 == 0: 
        result = result + i
    i += 1
print result

2550


while문을 활용하여 1부터 100사이의 짝수의 합을 계산하여 출력하시오.

while문에서는 range대신 c와 비슷한 조건문을 사용하여 범위를 지정해준 후에 위의 문제와 마찬가지로 %를 사용하여 짝수를 구하여, 구한값을 result에 계속 넣어주었다. {} 대신 탭으로 반복문 혹은 조건문이 작용하는 범위를 지정한다.

In [13]:
num = input('int : ')
print str(num)[::-1]

int : 9090
0909


사용자로 부터 임의의 정수를 입력받고, 해당 숫자를 역순으로 출력하는 프로그램을 작성하시오.

[실행 예]

  정수를 입력하세요: 3125
  숫자 역순은 5213 입니다.
  
input()을 사용하여 수를 받은 후에, 이를 str()을 사용하여 문자열로 바꾸었다. 그 후에 [::-1] 슬라이싱을 사용하여 역순으로 출력하였다. 9090이라는 수가 하나의 수이기 때문에 문자열로 바꾸어 뒤집는 것을 택했다.

In [1]:
num = input('int : ')
for i in range(1, num):
    check = 0
    for j in range(1,i):
        if i % j == 0:
            check += 1
    if check == 1 and i != 1:
        print i
     

int : 35
2
3
5
7
11
13
17
19
23
29
31


사용자로 부터 정수를 입력받아서 1부터 그 사이에 존재하는 소수 (Prime number)를 출력하는 파이썬 프로그램을 작성하시오.

for문에서 정수 범위를 받고, check는 나누어 떨어지는 수를 세기 위해 선언하였다. i는 i보다 작은 모든 양의 정수 j로 나누어 나머지를 검사하여 나누어떨어질때 check가 증가하였다. 
i = j일때 check는 1번 증가하기때문에 check 가 1인 i는 소수로 판정된다.

In [6]:
result = 0
for i in range(1,1000):
    if i % 3 == 0 or i % 5 == 0:
        result = result + i
print result

233168


[프로젝트 오일러 문제 1] 1000보다 작은 자연수 중에서 3또는 5의 배수를 모두 더한수
for문으로 범위를 지정하고, if문에서 3혹은 5의 배수일 때, 해당 값을 result에 더하도록 하였다.

In [18]:
fn = 0
result = 0
def fibo(n):
    if n <2 : return n
    else : return fibo(n-1) + fibo(n-2)

while(1):
    if(fibo(fn)>=4000000) : break
    if(fibo(fn)%2 ==0):
        result = result + fibo(fn)
    fn += 1
print '결과 : ' + str(result)

결과 : 4613732


[프로젝트 오일러 문제 2]피보나치 수열의 각 항은 바로 앞의 항 두 개를 더한 것이 됩니다. 1과 2로 시작하는 경우 이 수열은 아래와 같습니다.

1, 2, 3, 5, 8, 13, 21, 34, 55, 89, ... 

짝수이면서 4백만 이하인 모든 항을 더하면 얼마가 됩니까?


def를 이용해 피보나치 수를 정의하였다. while문은 4백만이하일때 계속 수행되며, 짝수인경우에 더해진다.

In [7]:
def primenumber(n):
    check = 0
    for i in range(1, n):
        if n % i == 0:
            check += 1
    if check == 1:
        return 1
    
temp = 0
i = 1
while(i < 13195):
    if(13195 % i == 0) and primenumber(i) == 1:
        temp = i
    i += 1
print temp


29


[프로젝트 오일러 문제 3]어떤 수를 소수의 곱으로만 나타내는 것을 소인수분해라 하고, 이 소수들을 그 수의 소인수라고 합니다.
예를 들면 13195의 소인수는 5, 7, 13, 29 입니다.

600851475143의 소인수 중에서 가장 큰 수를 구하세요.

구하기 위한 수가 너무 커서 컴퓨터가 계속 멈췄습니다. 그래서 예시에 있는 13195의 가장 큰 소인수를 출력시켰습니다. 소수를 찾는 함수를 조금 수정하여, 소수인경우 1을 반환하고, 13195를 나누어 떨어지는 동시에 소수이기때문에 1을 반환하는 수를 temp에 임시저장하였습니다. 여러 소인수가 나와도, 작은수부터 큰수로 덮어씌우기때문에 가장큰수가 temp에 저장되고 출력시킬때 원하는 값이 나왔습니다.
